### PR TITLE
[WIP] Abstract workings so that we can add other client(such as AzureDevOps)

### PR DESCRIPTION
**Pull request by MarcBruins**
First of all, i love the idea of NuKeeper and i feel very excited about it! This is my attempt to make the solution more generic so that we can easily add new client such as AzureDevOps or bitbucket.

The proposed solution structure will look like this:

- NuKeeper(Contains the command and the DI for registering the different clients)
- NuKeeper.Abstract (Contains the abstractionlayer for the Engine, the Configuration and some other shared code)
- NuKeeper.Github(Contains github implementation)
- NuKeeper.AzureDevOps(Contains AzureDevOps implementation)
- And the rest of the already existing parts

To let code function the same way for differenct clients i've extracted interfaces and used Githubs Octokit as a reference. I've used sharedcode where i can and made an implementation for Github specifically. So far everything builds but it doesn't work yet. I'm still working on getting the Github functionality back and would like to advance from there on to add AzureDevops.

The progress of this PR is:
- [x] Successful build with new project styles and abstraction layer
- [ ] Verify that all the same functionality that the package now has still works
- [ ] Update tests

This is WIP, please let me know what you think!

### CODE REVIEW

File: NuKeeper.Abstract/Engine/Packages/ExistingBranchFilter.cs [hunk 1]
```diff
@@ -1,29 +1,28 @@
...
 
         public async Task<bool> CanMakeBranchFor(PackageUpdateSet packageUpdateSet,
-            ForkData pushFork)
+            IForkData pushFork)
         {
             try
             {
                 var branchName = BranchNamer.MakeName(packageUpdateSet);
-                var githubBranch = await _gitHub.GetRepositoryBranch(pushFork.Owner, pushFork.Name, branchName);
+                var githubBranch = await _client.GetRepositoryBranch(pushFork.Owner, pushFork.Name, branchName).ConfigureAwait(false);
```
[the remark this answers is not in the document]
Comment (MarcBruins) [reply NuKeeper.Abstract/Engine/Packages/ExistingBranchFilter.cs]: Good point, i didn't know that. I'll change it back.

**Comment by AnthonySteele** [comment]
This is related to #336 

I haven't had time to look at this yet, I am away from home this weekend. Will do so when  can.
But it is in general a good idea to go in this direction of abstracting out github and generalising to other systems, hopefully in small steps.

File: NuKeeper.Abstract/NuKeeper.Abstract.csproj [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+        <TargetFramework>netstandard2.0</TargetFramework>
+    </PropertyGroup>
+
+    <ItemGroup>
+      <ProjectReference Include="..\NuKeeper.Inspection\NuKeeper.Inspection.csproj" />
+      <ProjectReference Include="..\NuKeeper.Update\NuKeeper.Update.csproj" />
+    </ItemGroup>
+
+    <ItemGroup>
+      <Reference Include="LibGit2Sharp, Version=0.26.0.0, Culture=neutral, PublicKeyToken=7cbde695407f0333">
```
Comment (AnthonySteele) [inline NuKeeper.Abstract/NuKeeper.Abstract.csproj]: This should be a `PackageReference` not a relative pathed dll reference. Issue with tooling?

File: NuKeeper.Abstract/NuKeeper.Abstract.csproj [hunk 3]
```diff
@@ -0,0 +1,18 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+    <PropertyGroup>
+        <TargetFramework>netstandard2.0</TargetFramework>
+    </PropertyGroup>
+
```
Comment (AnthonySteele) [inline NuKeeper.Abstract/NuKeeper.Abstract.csproj]: New libraries should also contain:
````
  <PropertyGroup>
    <CodeAnalysisRuleSet>..\CodeAnalysisRules.ruleset</CodeAnalysisRuleSet>
  </PropertyGroup>
````

**Comment by AnthonySteele** [comment]
So, a general design comment around things in `NuKeeper.Abstract` such as `interface IPullRequest` and `class GithubBranch : Branch, IBranch`

We're going to have to get this working well, and the plan (here)[https://github.com/NuKeeperDotNet/NuKeeper/issues/336#issuecomment-416992498] was to instead use Nukeeper's own types that "have only the few fields that we need". 

We are aware that the Octokit types are not ideal to use inside NuKeeper - the biggest symptom of that is tests where creating an Ocotkit object requires over 30 args to the constructor (!).  At some stage this might be changed to use github's graphQL API, and retrieve "only the few fields that we need". 

But it seems odd to me to use Interfaces and inheritance to solve this issue. Ideally I'd like to see all backends work on a common set of internal DTOs, that don't need interfaces or base classes. This does depend of the functionality being more or less equivalent so that common data types make sense.

**Comment by AnthonySteele** [comment]
Big PRs with lots of changes and also merge conflicts are my least favourite kind :(
There must be ways to slice this up into things that can be merged today, e.g. introduce just the "abstractions" project. Or just the "github" project.

I am happy with a PR that doesn't make a visible difference in itself; I think we have reasonable idea of where you are going.

**Comment by MarcBruins** [comment]
I would like to close this so that we can keep PR's smaller as @AnthonySteele mentions. That doesn't mean that i'm throwing away the progress that's been made. This PR can be kept as a reference towards a more abstract solution which enables more clients. I'll raise a issue with steps that i see so we can make a list of small PR's towards a good solution.

**Pull request closed without merging**